keps/sig-node/3857-rro-mounts/README.md

```diff
@@ -0,0 +1,1049 @@
+This might be a good place to talk about core concepts and how they relate.
+-->
+Constraints: needs runc >= 1.1 && kernel >= 5.12.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Risk: a manifest with `recursivelyReadOnly: Enabled` will not work on old distros (See "Constraints" above).
+
+Mitigation: set `recursivelyReadOnly: BestEffort`.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### Core API
+Add `RecursivelyReadOnly: (Disabled|BestEffort|Enabled)` to the [`VolumeMount`](https://github.com/kubernetes/kubernetes/blob/v1.26.1/pkg/apis/core/types.go#L1854-L1880) struct:
+- `Disabled` (default): always non-recursive
+- `BestEffort`: recursive if possible, otherwise non-recursive, without raising an error.
+   This value is translated to `Enabled` or `Disabled` in the `VolumeMountStatus` struct (see the "Diff" below).
+- `Enabled`: recursive if possible, otherwise raise an error (as long as the node components are aware of this feature).
+
+When `RecursivelyReadOnly` is set to any non-nil value, `ReadOnly` is validated to be `true`, and if `ReadOnly` is not specified,
+the default value of `ReadOnly` will be `true`.
+
+When a user sets `RecursivelyReadOnly` to `BestEffort` or `Enabled`, they should also set `MountPropagation` to `"None"` (default).
+Otherwise, a writable submount may be propagated from the host mount namespace.
+
+A pod manifest will look like this:
+```yaml
+spec:
+  volumes:
+    - name: foo
+      hostPath:
+        path: /mnt
+        type: Directory
+  containers:
+  - volumeMounts:
+    - mountPath: /mnt
+      name: foo
+      mountPropagation: None
+      # NEW
+      recursivelyReadOnly: BestEffort
+```
+
+Diff:
+```diff
+diff --git a/pkg/apis/core/types.go b/pkg/apis/core/types.go
+index 04ffd250b61..57ec182e3cc 100644
+--- a/pkg/apis/core/types.go
++++ b/pkg/apis/core/types.go
+@@ -1855,9 +1855,15 @@ type ContainerPort struct {
+ type VolumeMount struct {
+ 	// Required: This must match the Name of a Volume [above].
+ 	Name string
+-	// Optional: Defaults to false (read-write).
++	// Optional: Defaults to false (read-write), unless RecursivelyReadOnly is set to a non-nil value.
+ 	// +optional
+-	ReadOnly bool
++	ReadOnly *bool
++	// RecursivelyReadOnly specifies recursively-readonly mode.
++	// When RecursivelyReadOnly is set to a non-nil value, ReadOnly is implied to be true.
++	// Defaults to Disabled.
```

Comment:
If this Defaults to "Disabled" then it is always non-nil, which means the above statements about `ReadOnly` force that to be true.

Basically what I wrote above - does RO=false + RRO=Disabled cause an error?

Reply:
Reworded to:
```
When RecursivelyReadOnly is nil, ReadOnly is not validated to be true, and if ReadOnly is true,
it will be a non-recursive readonly mount.
```